tests/nn/data_parallel/test_fsdp_regnet.py

```diff
@@ -143,39 +235,67 @@ def _test_func(
     if fsdp_config:
         ddp = False
         assert isinstance(fsdp_config, dict), str(fsdp_config)
+        if fsdp_config["mixed_precision"]:
+            # To match DDP in AMP -O1, we need fp32 reduce scatter.
+            fsdp_config["fp32_reduce_scatter"] = True
 
     model = Model()
     model.load_state_dict(state_before)
     model = model.cuda()
 
+    class DummyScaler:
+        def scale(self, loss):
+            return loss
+
+        def step(self, optim):
+            optim.step()
+
+        def update(self):
+            pass
+
+    scaler = DummyScaler()
     if ddp:
         model = SyncBatchNorm.convert_sync_batchnorm(model)
-        model = DDP(model, device_ids=[rank])
+        model = DDP(model, device_ids=[rank], broadcast_buffers=True)
+        if ddp_mixed_precision:
+            scaler = GradScaler()
     else:
         # Note, different rank may wrap in different order due to different random
         # seeds. But results should be the same.
         if random.randint(0, 1) == 0:
-            print("auto_wrap_bn, then convert_sync_batchnorm")
-            model = auto_wrap_bn(model)
-            model = SyncBatchNorm.convert_sync_batchnorm(model)
+            print(f"auto_wrap_bn {fsdp_wrap_bn}, then convert_sync_batchnorm")
+            if fsdp_wrap_bn:
+                model = auto_wrap_bn(model, _single_rank_pg)
+            model = _bn_converter(model)
         else:
-            print("convert_sync_batchnorm, then auto_wrap_bn")
-            model = SyncBatchNorm.convert_sync_batchnorm(model)
-            model = auto_wrap_bn(model)
+            print(f"convert_sync_batchnorm, then auto_wrap_bn {fsdp_wrap_bn}")
+            model = _bn_converter(model)
+            if fsdp_wrap_bn:
+                model = auto_wrap_bn(model, _single_rank_pg)
         model = FSDP(model, **fsdp_config).cuda()
+        if fsdp_config["mixed_precision"]:
+            scaler = ShardedGradScaler()
+        # Print the model for verification.
+        if rank == 0:
+            print(model)
     optim = SGD(model.parameters(), lr=0.1)
+    loss_func = CrossEntropyLoss()
 
     for in_data in inputs[rank]:
         in_data = in_data.cuda()
         context = contextlib.suppress()
         if ddp and ddp_mixed_precision:
             in_data = in_data.half()
             context = torch.cuda.amp.autocast(enabled=True)
+        if not ddp and fsdp_config["mixed_precision"]:
+            context = torch.cuda.amp.autocast(enabled=True)
```

Comment:
just curious, when will this be true?

Reply:
this is when we use FSDP in mixed precision mode. So we need to turn on pytorch AMP.